### ISSUE TITLE
Big MPI data type test fails

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#12986]
On both my laptop and my dev workstation, the new test `mpi/create_mpi_datatype_01.mpirun=2.debug` fails. On my laptop, the assertion in question is 
```
 Assert(size64 == static_cast<MPI_Count>(n_bytes), ExcInternalError());
```
which fails because
```
(gdb) p size64
$9 = 4294967296
(gdb) p n_bytes
$10 = 8589934592
```

@tjhei: FYI. This was introduced in #12964.

**Comment by tjhei** [comment]
What kind of MPI implementation and version do you have?

**Comment by bangerth** [comment]
On my laptop, I use OpenMPI 2.1.1.

**Comment by tjhei** [comment]
I am surprised that you trigger the assert, as it is inside ``#      if DEAL_II_MPI_VERSION_GTE(3, 0)``. What MPI version does it state in your ``detailed.log``?

**Comment by bangerth** [comment]
I reinstalled my laptop yesterday, so I don't have that setup any more. But on my workstation I have MPI version 3.0, based on
```
#define OMPI_MAJOR_VERSION 1
#define OMPI_MINOR_VERSION 10
#define OMPI_RELEASE_VERSION 7

#define MPI_VERSION 3
#define MPI_SUBVERSION 0
```

**Comment by tjhei** [comment]
Ok, thanks. I can confirm that things are broken with OpenMPI (not only the test). 

That is a painful and annoying reminder that I should not assume things to work just based on one implementation (mpich in this case) on my machine.

**Comment by bangerth** [comment]
Bummer. It's annoying that none of the earlier MPI calls returns an error since clearly they must be doing something wrong.

Do you have a system on which you can test and check what's going wrong?

**Comment by tjhei** [comment]
So, OpenMPI 1.10.2 fails to create a datatype bigger than 2^32 (well, every call succeeds without error but the total size reported is wrong and I assume that the type is broken as well, not sure). You seem to have the same problem with 2.1.1.

I then checked OpenMPI 4.0.3 and it works correctly.

I am not sure what to do at this point. Note that all 3 versions are MPI 3 compliant.

**Comment by tjhei** [comment]
> I am not sure what to do at this point. Note that all 3 versions are MPI 3 compliant.

And it turns out that the types are indeed created correctly and you can use them (at least MPI_Send works just fine). It just happens that MPI_Type_size_x reports the wrong value. :-(

**Comment by tjhei** [comment]
> Can we just abort in that function based on the OpenMPI version? That's not particularly elegant, but avoids wrong results at least.

We could also just remove the Assert, because the MPI type seems to work correctly. The only thing broken on older OpenMPI is the test (when you are trying to inspect the size).

**Comment by tjhei** [comment]
In summary:
Intel MPI 19: works
MVAPICH2 2.3: works
MPICH 3.3.2: works
OpenMPI 4.03: works
OpenMPI 1.10.2, 2.1.1: test/assert fails (because MPI_Type_size_x reports the wrong value), but code runs correctly.

**Comment by bangerth** [comment]
I can confirm that OpenMPI 4.0.3 works for me as well.

Maybe I misunderstood: Are you saying that with older OpenMPI the actual data type is constructed as expected, and data sent with it is correctly transported, but that the `MPI_Type_size_x` function is the only one that has the bug? That is, for example, if we had a test that allocates 5GB of data, initializes it with (pseudo-)random numbers, broadcasts it to other processes with this type, that the data received on the other processes has 5GB of correct data?

**Comment by tjhei** [comment]
Correct! The test does exactly that as well (scroll to the bottom function).

**Comment by bangerth** [comment]
Ah yes, indeed. But like you say in your edit, it would be nice to test this for >2GB as well. (Separately, you only check the first and last element of the array, but not the values in between. That's probably ok, though now I'm paranoid :-) )

**Comment by bangerth** [comment]
@tjhei So what do we want to do about this?

**Comment by tjhei** [comment]
I am busy with other stuff this week, but I will get back to it soon. My first plan is to run some more tests to understand what works and what doesn't. Based on that, we can hopefully find something that works somewhat reliably. The whole thing is somewhat disappointing, though.

**Comment by bangerth** [comment]
Good plan, thanks for looking into it! (And yes, I feel you. Bugs in external software suck :-( )

**Comment by bangerth** [comment]
This seems to work now. Thanks, @tjhei !

**Comment by bangerth** [comment]
Scrap this. It works on my laptop, but that was due to the updated MPI implementation. It still fails on the workstation.
